core/trino-main/src/main/java/io/trino/operator/output/BytePositionsAppender.java

```diff
@@ -14,31 +14,239 @@
+    private static final Block NULL_VALUE_BLOCK = new ByteArrayBlock(1, Optional.of(new boolean[] {true}), new byte[1]);
+
+    @Nullable
+    private final BlockBuilderStatus blockBuilderStatus;
+    private boolean initialized;
+    private final int initialEntryCount;
+
+    private int positionCount;
+    private boolean hasNullValue;
+    private boolean hasNonNullValue;
+
+    // it is assumed that these arrays are the same length
+    private boolean[] valueIsNull = new boolean[0];
+    private byte[] values = new byte[0];
+
+    private long retainedSizeInBytes;
+
+    public BytePositionsAppender(@Nullable BlockBuilderStatus blockBuilderStatus, int expectedEntries)
+    {
+        this.blockBuilderStatus = blockBuilderStatus;
+        this.initialEntryCount = max(expectedEntries, 1);
+
+        updateDataSize();
+    }
+
+    @Override
+    public void appendTo(IntArrayList positions, Block block)
+    {
+        int[] positionArray = positions.elements();
+        int newPositionCount = positions.size();
+        ensureCapacity(positionCount + newPositionCount);
+        boolean hasNullValue = false;
+        boolean hasNonNullValue = false;
+
+        if (block.mayHaveNull()) {
+            for (int i = 0; i < newPositionCount; i++) {
+                int position = positionArray[i];
+                boolean isNull = block.isNull(position);
+                int positionIndex = positionCount + i;
+                if (isNull) {
+                    valueIsNull[positionIndex] = true;
+                    hasNullValue = true;
+                }
+                else {
+                    values[positionIndex] = block.getByte(position, 0);
+                    hasNonNullValue = true;
+                }
+            }
+            this.positionCount += newPositionCount;
+            this.hasNonNullValue |= hasNonNullValue;
+            this.hasNullValue |= hasNullValue;
+        }
+        else {
+            for (int i = 0; i < newPositionCount; i++) {
+                int position = positionArray[i];
+                values[positionCount + i] = block.getByte(position, 0);
+            }
+            positionCount += newPositionCount;
+            this.hasNonNullValue = true;
+        }
+
+        if (blockBuilderStatus != null) {
+            blockBuilderStatus.addBytes(ByteArrayBlock.SIZE_IN_BYTES_PER_POSITION * newPositionCount);
+        }
+    }
+
     @Override
-    public void appendTo(IntArrayList positions, Block block, BlockBuilder blockBuilder)
+    public Block build()
+    {
+        if (!hasNonNullValue) {
+            return new RunLengthEncodedBlock(NULL_VALUE_BLOCK, positionCount);
+        }
+        return new ByteArrayBlock(positionCount, hasNullValue ? Optional.of(valueIsNull) : Optional.empty(), values);
+    }
+
+    @Override
+    public void appendDictionary(IntArrayList positions, DictionaryBlock block)
```

Comment:
Make some sort of comment why this method is identical to appendTo. I guess this may be a generic comment  in `BlockTypeAwarePositionsAppender`